Made sure an external id can be associated to an already-registered key.

xavierlepretre opened the pull request:
Before, in both implementations of `IdentityService`, when calling `registerKey` with an external id, and the key was already existing, then the external id would be ignored.

Now, I changed so that if the key is known but has no external id, then the external id is associated with it.

This is my 2nd PR.

# PR Checklist:

- [x] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [ ] If you added public APIs, did you write the JavaDocs?
- [ ] If the changes are of interest to application developers, have you added them to the [changelog](https://docs.corda.net/head/changelog.html) (`/docs/source/changelog.rst`), and potentially the [release notes](https://docs.corda.net/head/release-notes.html) (`/docs/source/release-notes.rst`)?
- [ ] If you are contributing for the first time, please read the [contributor agreement](https://docs.corda.net/head/contributing.html) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://docs.corda.net/head/contributing.html#merging-the-changes-back-into-corda).

Thanks for your code, it's appreciated! :)

roastario committed:
add

roastario committed:
start work on liquibase schema for new checkpoints

roastario committed:
switch to new version of plugin

roastario committed:
change plugin version to force refresh

roastario committed:
try using pre-populated cache

roastario committed:
do not preallocate

roastario committed:
no daemon

roastario committed:
reduce cores per fork to increase parallelism

roastario committed:
Merge branch 'switch_to_local_k8s' into new_checkpoint_schema

# Conflicts:
#	Jenkinsfile

roastario committed:
address initial review comments

roastario committed:
add message to exceptions table

roastario committed:
even more aggresive cpu allocation

roastario committed:
:Merge branch 'switch_to_local_k8s' into new_checkpoint_schema

roastario committed:
Merge branch 'release/os/4.4' of github.com:corda/corda into new_checkpoint_schema

roastario committed:
start adding hibernate entities

roastario committed:
flesh out entities for new checkpointing

roastario committed:
working node startup with new tables and entities

roastario committed:
persist some nonsense data to see if it breaks the statemachine

roastario committed:
Merge branch 'release/os/4.4' of github.com:corda/corda into new_checkpoint_schema

roastario committed:
address dan review comments

roastario committed:
address dan review comments pt2

roastario committed:
Merge pull request #5892 from corda/new_checkpoint_schema

New checkpoint schema

roastario committed:
Merge remote-tracking branch 'origin/feature/checkpoint_table_improvements' into feature/checkpoint_table_improvements

williamvigorr3 committed:
CORDA-3432 update structure of checkpoint class (#5983)

* Split StateMachine State into 2 classes

The idea is this better reflects the database structure. Added a
few helper methods to copy and update state.

* Doc + Improve Checkpoint API

* Rename methods to be more clear

williamvigorr3 committed:
CORDA-3597 add missed data to checkpoint class (#5995)

* Add extra fields to StateMachineState

* Move structures into checkpoint as this is a more natural place.

williamvigorr3 committed:
CORDA-3597 Replace old Checkpoint table with new one. (#5992)

* Replace old Checkpoint table with new one.

Adds some of the new fields into the table where needed (I have
guessed this stuff but we can update it as we go along).

* Fix database constraints + name table correctly opps.

* Fixed typos in Liquidbase script

Also corrected constraints and added missed fields in hibernate
checkpoint class and liquibase scripts.

* Update CheckpointStorage to pass in serialization context.

This is cleaner than passing both the checkpoint and the
serialized checkpoint into the methods. Also fixed
CordaPersistanceServiceTests which I accidentally broke.

* Fix detekt problem

* Revert "Update CheckpointStorage to pass in serialization context."

This reverts commit b71e78f20274ab0f5b3cf3fda1451ae2bd7a6797.

* Fix test broken by reverting commit

* CORDA-3597 Update metadata join, timestamp columns and serialization

- Change the metadata join to the checkpoints table to use
  `invocation_id` instead of `flow_id`. There were issues joining
  between the tables because `flow_id` was not the primary key of the
  metadata table. Switching over to `invocation_id` has at least allowed
  us to bypass this issue. The information about the `invocation_id` is
  stored in the `Checkpoint` class which makes it simple to save at
  runtime.
- Some of timestamp columns were nullable when they should always be
  populated, the nullable flags have now been removed.
- Previously the whole checkpoint was being serialized and stored into
  the `checkpoints_blob.checkpoint` column. This meant duplicated saving
  as the `flow_state` was contained in this object. Only the
  `CheckpointState` property of `Checkpoint` is now being serialized and
  saved to this field. Furthermore, it now uses the default
  `STORAGE_CONTEXT` serialization (AMQP) instead of Kryo (which is only
  used for serializing the `flow_state` / flow stack).
- The checkpoint database performance metrics recording has been
  abstracted to its own class.

* CORDA-3597 Make metadata join non optional

Remove the nullable declaration on the metadata field of
`DBFlowCheckpoint`

* CORDA-3597 Rename `node_checkpoints_blobs` to `node_checkpoint_blobs`

* CORDA-3597 Update some kdocs

Co-authored-by: Dan Newton <danknewton@hotmail.com>

lankydan committed:
NOTICK Do not cascade checkpoint tables (#6040)

Do not cascade updates to checkpoint error and result tables to hopefully
improve database performance moving forward. Because the joined tables
are no longer being updated by updating the main `DBFlowCheckpoint` entity,
they must be created/updated/deleted manually.

The checkpoint blobs still cascade as they pretty much always evolve in 
tandem with the main checkpoint table.

williamvigorr3 committed:
CORDA-3600 Add flowIO request to checkpoint (#6017)

* Update Checkpoint DB to update flow io request

* Modify flow monitor to update Checkpoint DB with waiting flows

This happens periodically.

* Refactored code to avoid looping twice and updated tests

* Fix tests after rebasing

* Fix MR comments (non-functional refactor of tests + FlowMonitor).

* Made visible for testing method private in DBCheckpointStorage

This is not needed anymore.

* Explicity check if ioRequestType has changed in update method

* Fix shadowing warning

* Import non deprecated Assert into test

* Use AssertEquals not assert in test

* Address more comments (minor refactor) of DBCheckpointStorage

* Minor fix use it instead of referencing object explicitly

* Add null check to DBCheckpointStorage

* Revert changes to Flow Monitor.

We will instead store the information in the main thread of the
state machine.

* Remove now uneeded API and make statemachine update ioRequest

* Add Integration Test to check statemachine updates DB on Recieve

* Use simpleName in checkpoint storage instead of class.

Hibernate was previously resetting the class field this is now
set to null (when getting checkpoint form DB) and a new method
for getting back the simple name as a string.

* Update StateMachineState to store simple name.

* Fix after rebase broke stuff + renamed test

* Fix Detekt issue

* Remove uneeded null assertion

lankydan committed:
Merge branch 'release/os/4.5' into dan/merge-4.5-into-feature-branch

# Conflicts:
#	node/src/main/kotlin/net/corda/node/services/statemachine/transitions/StartedFlowTransition.kt

williamvigorr3 committed:
CORDA-3603 Save completed flow information (#6034)

When a flow is finished do not delete the checkpoint from the DB.

Instead, the FlowStatus is marked as Completed in the DB.

Updated numerous tests which relied on the flow being removed
when finished.

kyriathar-r3 committed:
CORDA-3598 Set Checkpoint.status to RUNNABLE (#6019)

* Set/ Reset Checkpoint.status to RUNNABLE after when suspending

* Removing/ Moving comment as it makes no longer sense to be there since, we now always create a new Checkpoint object in SingleThreadedStateMachineManager.createFlowFromCheckpoint through tryDeserializeCheckpoint

* Set -in memory- Checkpoint.status to RUNNABLE when a flow is retrying from Checkpoint

lankydan committed:
Merge branch 'feature/checkpoint_table_improvements' into dan/merge-4.5-into-feature-branch

# Conflicts:
#	node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt

williamvigorr3 committed:
CORDA-3599 Add progress tracker information to checkpoint (#6063)

* Add progress tracker information to checkpoint

The checkpoint Datebase is updated when the statemachine suspends
with the progress trackers current step name. This is truncated if
it is longer than the Database column.

* Minor rename in statemachine for clarity

kyriathar-r3 committed:
CORDA-3655 Do not run COMPLETED, FAILED or KILLED flows on node startup (#6070)

* CheckpointStorage.getAllCheckpoints will not fetch COMPLETED, FAILED and KILLED flows by default

* Rename getAllCheckpoints to getAllRunnableCheckpoints for clarity

* Fix Detekt issue

* Rename getAllRunnableCheckpoints to getRunnableCheckpoints

* Minor kdoc update

* Bring back in CheckpointStorage.getAllCheckpoints to co-exist with getRunnableCheckpoints

lankydan committed:
Merge branch 'feature/checkpoint_table_improvements' into dan/merge-4.5-into-feature-branch

lankydan committed:
Merge pull request #6058 from corda/dan/merge-4.5-into-feature-branch

Dan/merge 4.5 into feature branch

lankydan committed:
CORDA-3596 Record flow metadata (#6067)

* CORDA-3596 Record flow metadata

Record flow metadata during the zero'th checkpoint that occurs before
calling the flow's `call` function.

This required adding an RPC call's arguments to the `InvocationContext`
that gets created. These arguments are then accessible within the
statemachine and from the `Checkpoint` class. The arguments are then
extracted when recording a flow's metadata inside of
`DBCheckpointStorage`.

Updated the size of the started by column to 128 since it was not long
enough to hold the fully qualified class of a service that started a
flow.

* CORDA-3596 Remove arguments from in-memory checkpoint

When executing a flows first real suspend (from flow code) the arguments
 contained in the `InvocationContext` are removed. This saves holding
 these arguments for the whole lifecyle of a flow.

* CORDA-3596 Increase `cordapp_name` column to 128

* CORDA-3596 Join metadata by `flow_id`

Due to changes in where metadata is recorded, there is no need for
having `invocation_id` as the metadata table's primary key. The
`flow_id` is now the primary key of the table and is used to join to the
 main checkpoints table.

The `invocation_id` has been removed from the checkpoints table since it
 is not needed for the join anymore.

* CORDA-3596 Remove `received_time` from metadata table

* CORDA-3596 Remove unused `StartReason` enum

* CORDA-3596 Simple `DBCheckpointStorageTests` for metadata

* CORDA-3596 Truncate really long flow names

lankydan committed:
NOTICK Rename `rpc_user` column to `started_by` in metadata table (#6081)

kyriathar-r3 committed:
Merge branch 'release/os/4.5' into os_4.5-feature_checkpoint_table_improvements-merge

# Conflicts:
#	node/src/main/kotlin/net/corda/node/services/statemachine/transitions/StartedFlowTransition.kt

lankydan committed:
Merge pull request #6084 from corda/os_4.5-feature_checkpoint_table_improvements-merge

[NOTICK] OS 4.5 to feature/checkpoint_table_improvements merge

lankydan committed:
CORDA-3601 Record a flow's finish time (#6079)

* CORDA-3601 Record a flow's finish time

Record a flow's finish time by updating its metadata record. It is set
in `updateCheckpoint` by checking the status of the checkpoint. If it is
 `COMPLETED` it will set the `finishInstant` on the metadata object and
 update it.

* CORDA-3601 Record flow finish time for all finished statuses

Update the flow finish time for the following statuses:

- COMPLETED
- KILLED
- FAILED

* CORDA-3601 Use platform clock in `DBCheckpointStorage`

kyriathar-r3 committed:
Merge branch 'release/os/4.5' into os_4.5-feature_checkpoint_table_improvements-merge

lankydan committed:
Merge pull request #6105 from corda/os_4.5-feature_checkpoint_table_improvements-merge

[NOTICK] OS 4.5 to feature/checkpoint_table_improvements merge

williamvigorr3 committed:
CORDA-3491 Remove the flow state when a flow finishes (#6083)

Added a new field Completed to the in-memory object FlowState.

FlowState.Completed is corresponds to flow_state=Null in the DB.

This change will save disk space.

kyriathar-r3 committed:
CORDA-3604 Store failed and hospitalised errors along with corresponding statuses (#6061)

Flows that are kept for overnight observation:

- Save their Checkpoint.status as 'HOSPITALIZED' in the database
- Save the error that caused the hospitalization in the database

A new Event was added for this reason. Whenever the hospital determines
a flow for hospitalization, it adds this Event in the flow's fiber queue.
When processed it creates a new DB transaction, stores the checkpoint status along with
the error, and it adds a 'FlowContinuation.ProcessEvents' continuation so that the fiber keeps
processing events (effectively since there are no more events in the fiber's channel, the fiber will suspend).

Flows that error:

- Their checkpoints are kept in the database
- Save their Checkpoint.status as 'FAILED'
- Save the error that caused the error in the database

Upon erroring, the flow's Checkpoint.status gets updated('FAILED') and the checkpoint is stored
in the database instead of getting removed. The flow then propagates the error to counterparties,
sets its future with the error and gets removed from memory.

lankydan committed:
CORDA-3691 delete checkpoint record when complete (#6134)

* CORDA-3691 Delete checkpoint when flow finishes

The checkpoint and its related records in joined tables should be deleted
when a flow finishes.

Keeping these flows around will be completed in the future.

* CORDA-3691 Ignore some flow metadata tests

Ignore tests around recording the finish time of flow metadata records
since we are not currently keeping COMPLETED flows in the database.

lankydan committed:
Merge branch 'release/os/4.6' into dan/4.6-into-checkpoint-feature-branch-2020-05-05

# Conflicts:
#	node/src/main/kotlin/net/corda/node/services/statemachine/Event.kt
#	node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt
#	node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt
#	node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt

lankydan committed:
NOTICK Fix `KilledFlowTransition` after merge

lankydan committed:
NOTICK Remove unused imports

lankydan committed:
NOTICK Fix kill flow tests due to storing failed flows

Failed flows are stored after the checkpoint table rework. This meant
that some of the asserts in `FlowIsKilledTest` and `KillFlowTest` were
wrong.

lankydan committed:
NOTICK Change platform version assert in `FlowMetadataRecordingTest`

lankydan committed:
Merge pull request #6220 from corda/dan/4.6-into-checkpoint-feature-branch-2020-05-05

Merge 4.6 into checkpoint feature branch 2020-05-05

xavierlepretre committed:
Made sure an external id can be associated to an already-registered key.

xavierlepretre committed:
Made sure an external id can be associated to an already-registered key.

xavierlepretre committed:
Added some tests to the InMemoryIdentityService too.

xavierlepretre committed:
Added some tests to the InMemoryIdentityService too.

roastario commented:
thanks for this - it looks good to me, just going to have a talk with @roger3cev and a few others tomorrow before merging.

roger-that-dev commented:
Thanks for the PR. What's the use-case where you need to assign an external ID to an already existing key? Surely when you generate a public key you know which account it's for, there and then?

Schife committed:
INFRA-331 Improving Jenkins PR build performance (#6252)

* INFRA-331 increasing shard number and reintroducing pre allocation

* INFRA-331 removing preallocation

* INFRA-331 reintroducing preallocation as it cuts 10 mins from the build

* INFRA-331 test performance with 20 pods

* INFRA-331 test performance with 10 pods and preallocation

* INFRA-331 more tests on 15 pods + prealloc

roger-that-dev commented:
Also, as I understand it, this is being changed for the training? Is that really necessary? @mnesbit - thoughts? This impacts kernel.

alch3mistdev commented:
This is a particular scenario utilising the requested behaviour.

https://github.com/corda/corda-training-code/blob/3ad90b19b2bb0febc5d996b769cc9c63c3cbcb81/040-accounts-lib/workflows/src/test/java/com/template/account/AccountVaultCourseExercise.java#L99

@xavierlepretre feel free to comment or add any context if needed.

xavierlepretre commented:
The use case pasted above is in a unit test demonstrating the association, or absence thereof, between an id and a pub key. 
The post hoc setting of an id to the pub key is for the benefit of a node that did not create the key in the first place but instead received it by one of many ways. Say:

- Alice creates key1 and key2 with account of id1
- Alice launches a flow that uses `SyncKeyMappingFlow` to send key1 over to Bob. It works. There is no id1 associated to key1 because this is not what `SyncKeyMappingFlow` does.
- Later Bob launches a flow that use `SyncKeyMappingFlow` to send key1 over to Carol. It works. There is no id1 associated...
- Later, Alice launches another flow that uses `ShareStateAndSyncAccountsFlow` to send the account id1 info plus key1 and key2 over to Carol. Carol is surprised that her vault now has 2 pub keys supposedly from the account Alice sent, but only key2 has the corresponding id1.

What do we tell Carol? Sorry mate, Alice should not have launched a flow with Bob that used `SyncKeyMappingFlow`. Better luck next time.

This PR makes sure that key1 is also associated with id1 whatever the previous choices of Alice.

roger-that-dev commented:
OK, I get your point. I'll leave it up to @mnesbit as it's his call. Cheers

JamesHR3 committed:
EG-2116 - Add aliases for version attribute errors (#6260)

nikinagy committed:
empty list checks (#6262)

Commit:
CORDA-3805: Make custom migration scripts independent on the latest PersistentIdentityService schema

williamvigorr3 committed:
CORDA-3490 Add option to start node without starting checkpointed flows (#6136)

Added command-line option: `--pause-all-flows` to the Node to control this.
This mode causes all checkpoints to be set to status PAUSED when the
state machine starts up (in StartMode.Safe mode). 

Changed the state machine so that PAUSED checkpoints are loaded into
memory (the checkpoint is deserialised but the flow state is left serialised)
but not started.

Messages from peers are queued whilst the flow is paused and processed
once the flow is resumed.

lankydan committed:
NOTICK Fix timing issue in `FlowIsKilledTest` (#6264)

Also remove `Configurator` that seems to fail on some agents.

nikinagy committed:
ENT-5278 - add missed checks for empty lists (#6261)

* add missed checks for empty lists

* adding missing brackets

* adding missing timeout

Schife committed:
INFRA-350 add suffix to the jdk 11 image to avoid overwrite (#6268)

relyafi committed:
[NOTICK] Remove obsolete on-demand smoke test Jenkins configuration (#6265)

Schife committed:
INFRA-350 add suffix to the jdk 11 image to avoid overwrite (#6269)

nikinagy committed:
Merge branch 'release/os/4.5' into nnagy-os-4.5-os-4.6-20200520

relyafi committed:
[INFRA-351] Jenkinsfile for code check jobs (#6272)

woggioni committed:
improved interactive shell flow lookup logic (#6271)

the current logic when a user type
```
flow start SomeFlow
```
is to search for the string `SomeFlow` in all registered flow names, if there is at exactly flow that ends with that string, it will be selected,
if there is more than one, the first in aphabetical order will be selected. If there are multiple flows that contains the string but no one that ends with it, it is considered ambiguous,
if there is exactly one containing the string it is instead selected (it doesn't matter whether it ends or not with it). All other cases are considered errors.
The goal of this change is to address the case where the node contains

```
net.corda.cordappp.AnotherSomeFlow
net.corda.cordappp.ImprovedSomeFlow
net.corda.cordappp.SomeFlow
```

typing
```
flow start SomeFlow
```
currently results in running `net.corda.cordappp.AnotherSomeFlow` because it comes first in alphabetical order, while `net.corda.cordappp.SomeFlow` is a better matches because the input substring
matches a bigger portion of the full flow name

Commit:
CORDA-3800: Flaky ProtonWrapperTests test investigation (#6258)

rick-r3 committed:
Merge pull request #6270 from corda/nnagy-os-4.5-os-4.6-20200520

Nnagy os 4.5 os 4.6 20200520

nikinagy committed:
CORDA-3587 - adding kdocs for current behaviour of VaultQueryCriteria (#6242)

* adding kdocs for current behaviour of VaultQueryCriteria

* improving the kdocs

* address PR comments

kyriathar-r3 committed:
ENT-5339 Failing tests against Oracle in VaultObserverExceptionTest (#6275)

* Fix erroneous sql statement for oracle; It was failing tests with 'ORA-00933: SQL command not properly ended'

* Fixed flaky test; it didn't wait for counter party flow to get hospitalized as the test implied

nikinagy committed:
Merge branch 'release/os/4.3' into nnagy-os-4.3-os-4.4-20200522

mnesbit committed:
Merge pull request #6281 from corda/nnagy-os-4.3-os-4.4-20200522

Nnagy os 4.3 os 4.4 20200522

nikinagy committed:
Merge branch 'release/os/4.4' into nnagy-os-4.4-os-4.5-20200526

# Conflicts:
#	node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt

lankydan committed:
ENT-4627 Log transaction context missing exception (#6278)

When a flow is missing its database transaction the _real_ stacktrace
of the exception is missing in the logs. This is due to the normal error
handling path way does not work due the transaction being missing. When
it goes to process the next error event (due to the original transaction
context missing error) it will fail for the same error as it needs the
context to be able to process the error event.

The extra logging should help diagnose future errors.

relyafi committed:
[INFRA-352] Artifactory publication in Jenkins (#6276)

* [INFRA-352] Artifactory publication in Jenkins

* Address review comments

chrisr3 committed:
CORDA-3750: Reimplement Corda's Crypto object for use inside the sandbox. (#6193)

* CORDA-3750: Use hand-written sandbox Crypto object that delegates to the node.

* CORDA-3750: Add integration test for deterministic CashIssueAndPayment flow.

* Tidy up generics for Array instances.

* Upgrade to DJVM 1.1-RC04.

Commit:
CORDA-3818: Synchronize OS implementation of PublicKeyToOwningIdentityCache with CE

chrisr3 committed:
Merge commit '6ebc6e9b16575a7afbb781b0d93a8f04b3affba5' into chrisr3-45-merge

mnesbit committed:
Merge pull request #6284 from corda/nnagy-os-4.4-os-4.5-20200526

Nnagy os 4.4 os 4.5 20200526

adelel1 committed:
CORDA-3755: Backport AttachmentURLStreamHandlerFactory memory leak (#6274)

* CORDA-3755: Switched attachments map to a WeakHashMap (#6214)

* CORDA-3772: Now specify source and target of 8 when compiling contract classes.

* CORDA-3651: addManifest now uses separate files for reading and writing. (#6026)

* CORDA-3651: addManifest now uses separate files for reading and writing.

* CORDA-3651: The jar scanning loader now closes itsself.

Co-authored-by: Adel El-Beik <adelel-beik@19LDN-MAC108.local>

Co-authored-by: Adel El-Beik <adelel-beik@19LDN-MAC108.local>

Commit:
Merge pull request #6288 from corda/chrisr3-45-merge

Merge from OS 4.5 up to 6ebc6e9

kyriathar-r3 committed:
CORDA-3608 Performance tune the new checkpoint schema (#6277)

Performance tuning of the new checkpoint schema;

- Checkpoint tables are now using `flowId` as join keys. 
- Indexes consist of a PK's index on `node_checkpoints(flow_id)` and then unique indexes on `node_checkpoint_blobs(flow_id)` and `node_flow_metadata(flow_id)`.
- Serialization of `checkpointState` is being done with `CHECKPOINT_CONTEXT` so that we can have compression. This is needed when messages get passed into `checkpointState.sessions` therefore `checkpointState` grows in size upon serialized and 
saved into the database.

* Deserialize checkpointState with CHECKPOINT_CONTEXT

* Align tests with schema update; We cannot add and update a checkpoint in the same session now, ends up with hibernate complaining: two different objects with same identifier

* Fix indentation and format

* Ignore tests that assert DBFlowResult or DBFlowException

* Set DBFlowCheckpoint.blob to null whenever the flow errors or hospitalizes; this way we save an extra SELECT in such cases;

* Fix test; cleared Hibernate session, it would fail at checkpoint with 'org.hibernate.NonUniqueObjectException'

* Changing VARCHAR to NVARCHAR

* Rename v17 liquibase scripts to v19 to resolve collision with ENT v17 scripts

alch3mistdev commented:
@rekalov Could you please take a look at this when you have a moment - to make sure it's compatible with any refactoring you're doing? thx

nikinagy committed:
Merge branch 'release/os/4.5' into nnagy-os-4.5-os-4.6-20200528

Commit:
Merge pull request #6285 from corda/denis/CORDA-3818-sync-identity-service

CORDA-3818: Synchronize OS implementation of PublicKeyToOwningIdentityCache with CE

rekalov commented:
I doesn't conflict with cert rotation work which I'm doing at the moment. Just check that changes are still mergeable after https://github.com/corda/corda/pull/6285.

As for the change itself, I don't know much about implementation, so can't comment if it should/can be changed in the node or in the flows.

One side note: it's not a thread safe, in the sense that two concurrent `registerKey` updating `externalId` may result in unpredictable behavior. Though, existing implementation of `registerKey` may also have such issues.

mnesbit commented on node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt:
```diff
@@ -131,13 +131,19 @@ class InMemoryIdentityService(
+            if (existingExtId == null) {
+                registerKeyToExternalId(publicKey, externalId)
+            } else {
+                if (existingExtId != externalId) {
```
what is meant to go in this branch of if tree?

roger-that-dev replied:
nothing, probably a bad merge

xavierlepretre replied:
It mimics the absence of a throw like on lines 135-136 of this very file.
The same `if` structures, but with throws, are [found](https://github.com/corda/corda/pull/6256/files/776e27b537f895417a28700cd1cb05b6f46ba078#diff-4242a6f0ed90c990808e6d0126ce9281R402-R416) in the `PersistentIdentityService` file.

mnesbit replied:
our coding standard is no dead, or commented out code so please remove empty if statement

mnesbit replied:
remove empty logic

xavierlepretre replied:
Should I also remove the empty block on lines 135-136?

mnesbit replied:
yes just required logic blocks should be there. The others just confuse and risk merge problems at some point

roger-that-dev commented:
@rekalov with the existing implementation you couldn't update a key which was already registered. Instead you could only ever add a key and assign it to an external ID.

rekalov commented:
@roger3cev , are you saying that it's not possible to have two concurrent `registerKey` calls on the same  key?

roger-that-dev commented:
@rekalov In the original version, you could only ever create a key and assign it to an ID in the same database transaction. So, from from the perspective of another flow... the key either exists and is maybe assigned to a UUID, or it doesn't exist. But a key that is not assigned to an ID, cannot be assigned one after it was created as the api didn't support it. So... even though it's highly unlikely two flows would choose to concurrently register the same key at the same time, even if they did there would be no strange side effects. The database would serialise the transactions and the second call to register key would be idempotent. I haven't looked at the new code, though.

edit: the above assumes flows don't run in parallel. If they do run in parallel, two flows looking to register the same key shouldn't cause any conflict, assuming they are registering it to the same party, which they should be!

rekalov commented:
Even registering the same key (against the same party) may result in duplicated DB calls. 
It's fully OK for `keyToName` and `hashToKey` which are implemented using `AppendOnlyPersistentMap` and should be safe.

However, it's different for `registerKeyToExternalId()` which uses pure Hibernate `session.persist()` under the hood. Most likely, the second call will fail.

roger-that-dev commented:
@rekalov ah that's true... even still... a _very_ unlikely scenario.

rekalov commented:
Thanks @roger3cev , understood.

roger-that-dev commented:
@rekalov that doesn't mean we shouldn't fix it though :)

rick-r3 committed:
Merge pull request #6292 from corda/nnagy-os-4.5-os-4.6-20200528

Nnagy os 4.5 os 4.6 20200528

relyafi committed:
[INFRA-352] Artifactory publication in Jenkins (#6276) (#6290)

* [INFRA-352] Artifactory publication in Jenkins

* Address review comments

kyriathar-r3 committed:
CORDA-3725 Fix SQL deadlocks coming from soft locking states (#6287)

Adding to the query -explicitly- the flow's locked states resolved the SQL Deadlocks in SQL server. The SQL Deadlocks would come up  from `softLockRelease` when only the `lockId` was passed in as argument. In that case the query optimizer would use `lock_id_idx(lock_id, state_status)` to search and update entries in `VAULT_STATES` table. 

However, all rest of the queries would follow the opposite direction meaning they would use PK's `index(output_index, transaction_id)` but they would also update the `lock_id` column and therefore the `lock_id_idx` as well, because `lock_id` is a part of it. That was causing a circular locking among the different transactions (SQL processes) within the database.

To resolve this, whenever a flow attempts to reserve soft locks using their flow id (remember the flow id is always the flow id for the very first flow in the flow stack), we then save these states to the fiber. Then, upon releasing soft locks the fiber passes that set to the release soft locks query. That way the database query optimizer will use the primary key index of VAULT_STATES table, instead of lock_id_idx in order to search rows to update. That way the query will be aligned with the rest of the queries that are following that route as well (i.e. making use of the primary key), and therefore its locking order of resources within the database will be aligned with the rest queries' locking orders (solving SQL deadlocks).

* Fixed SQL deadlocks caused from softLockRelease, by saving locked states per fiber; NodeVaultService.softLockRelease query then uses VAULT_STATES PK index instead of lock_id_idx

Speed up SQL server by breaking down queries with > 16 elements in their IN clause, into sub queries with 16 elements max in their IN clauses

* Allow softLockedStates to remove states

* Add to softLockedStates only states soft locked under our flow id

* Fix softLockRelease not to take into account flowStateMachineImpl.softLockedStates when using lockId != ourFlowId

* Moved CriteriaBuilder.executeUpdate at the bottom of the file

relyafi committed:
[INFRA-351] Jenkinsfile for code check jobs (#6294)

rick-r3 committed:
Merge pull request #6295 from corda/feature/checkpoint_table_improvements

CORDA-3432 Feature/checkpoint table improvements

nikinagy committed:
throwing clearer error message when not supported 301 response code is used (#6296)

relyafi committed:
Merge remote-tracking branch 'origin/release/os/4.4' into ramzi/44-45-merge

mnesbit committed:
Merge pull request #6298 from corda/ramzi/44-45-merge

OS 4.4 -> 4.5 merge 20200601

Vengi58 committed:
CORDA-3176 Inefficient query generated on vault queries with custom p… (#6241)

* CORDA-3176 Inefficient query generated on vault queries with custom paging

a check added to avoid self joins
local postgres db tests were executed on large volume test data (50k states) to ensure that the DB optimizes the suspicious query and so it does not cuase performance issues.
A test added to check that a pagination query on large volume of sorted data is completed in a reasonable time

* foreach detekt fix

JamesHR3 committed:
[EG-2148] Small tweak to an error message (#6263)

JamesHR3 committed:
[EG-2225] Prevent extra directories being created when relative base directories are specified (#6282)

Don't create an extra directory if a relative base path specified

williamvigorr3 committed:
NOTICK Restrict to shell commands (#6303)

Remove shell command for pausing flows

kyriathar-r3 committed:
NOTICK Fix schema validation error for flow parameters (#6304)

Adding 'corda-blob' type, fixed 'Schema-validation: wrong column type'

relyafi committed:
[INFRA-368] Add deployNodes and install gradle tasks (#6302)

lankydan committed:
NOTICK Correct `StatemachineGeneralErrorHandlingtest` (#6306)

Commit:
Merge pull request #6266 from corda/denis/CORDA-3805-custom-migration-scripts

CORDA-3805: cut dependency from PersistentIdentityService for custom migration scripts

Schife committed:
INFRA-381 Removing sonarqube (#6312)

nikinagy committed:
ENT-4064, ENT-4608 - checking for unsigned cordapps in prod mode and checking the PV (#6291)

* checking for unsigned cordapps in prod mode and shutting down if it the contracts are not signed

* inheriting from CordaRuntimeException instead of Exception

* had the same tests twice, removed the shutdown for minimumplatformversion, modified some of the tests

* shut down when we get invalid platformversion and modified the test according to this

* making the message more accurate

Schife committed:
Merge branch 'release/os/4.5' of https://github.com/corda/corda into razvan-os-4.5-to-4.6-merge

Schife committed:
INFRA-381 correctly merged jenkinsfile

lankydan committed:
Merge pull request #6314 from corda/razvan-os-4.5-to-4.6-merge

NOTICK OS 4.5 -> OS 4.6 merge

Schife committed:
INFRA-348 JDK 11 Compatibility job migration (#6311)

relyafi committed:
Merge remote-tracking branch 'origin/release/os/4.4' into ramzi/44-45-merge

wzur-r3 committed:
Merge pull request #6317 from corda/ramzi/44-45-merge

OS 4.4 -> 4.5 merge 20200605

relyafi committed:
Merge remote-tracking branch 'origin/release/os/4.5' into ramzi/45-46-merge

lankydan committed:
Merge pull request #6319 from corda/ramzi/45-46-merge

[NOTICK] OS 4.5 -> 4.6 merge 20200608

Commit:
ENT-5379 Reconnecting RPC fixed to recognize shutdown calls and break reconnect attempts (#6316)

blsemo committed:
NOTICK Remove example code (#6327)

* Remove example code

blsemo committed:
Merge commit '550b446' into christians/NOTICK-merge-OS-4.5-4.6

blsemo committed:
Merge pull request #6328 from corda/christians/NOTICK-merge-OS-4.5-4.6

NOTICK merge removal of example-code

andrehgl committed:
update docker version (#6215)

corda-zulu-java1.8-4.4

JamesHR3 committed:
EG-2564 - Move printed error to logger (#6323)

JamesHR3 committed:
EG-2225 - Create log directory in correct place with verbose flag set (#6321)

* Ensure logs directory is written to correct location
* Remove a superfluous set of log path property
* Add a unit test to catch bad log paths
* Address detekt issues

Schife committed:
INFRA-242 Adding github action to check PR titles (#6299)

* INFRA-242 Adding github action to check PR titles

* INFRA-242 rename yml file

fowlerrr committed:
NOTICK: Fix IRSDemo code to wait for a response when updating day (#6329)

JamesHR3 committed:
Merge branch 'release/os/4.5' into jamesh/os-4.5-4.6-merge-11062020

josephzunigadaly committed:
CORDA-3837: When delivering a lifecycle event, copy the database context into the lifecycle event thread (#6331)

JamesHR3 committed:
NOTICK - Ignore a potentially dodgy test (#6336)

lankydan committed:
Merge pull request #6332 from corda/jamesh/os-4.5-4.6-merge-11062020

NOTICK OS 4.5 -> OS 4.6 merge 11/06/2020

chrisr3 committed:
CORDA-3738: Upgrade to DJVM 1.1. (#6338)

chrisr3 committed:
Merge commit 'ef00fa1388db37e155ab8cfed3763c14801f8aa9' into chrisr3-45-merge

Schife committed:
INFRA-387 preallocation and deployNodes migration to Code Checks (#6339)

* preallocate for slow integration tests as well

Schife committed:
Merge branch 'release/os/4.4' of https://github.com/corda/corda into raz/os44to45

roastario committed:
Make Checkpoint classes data classes (#6342)

* Make Checkpoint classes data classes

* tidy up null-checks for array equality

blsemo committed:
Merge pull request #6341 from corda/chrisr3-45-merge

NOTICK: Merge from OS 4.5 up to ef00fa1.

JamesHR3 committed:
EG-2684 Ensure original message is logged with old-style error codes (#6340)

* [EG-2684] Ensure original message is logged with old-style error codes

* [EG-2684] Address detekt issues

wzur-r3 committed:
Merge pull request #6346 from corda/raz/os44to45

NOTICK: MERGE OS 4.4 -> OS 4.5

Schife committed:
Merge branches 'release/os/4.5' and 'release/os/4.6' of https://github.com/corda/corda into raz/os45to46

 Conflicts:
	Jenkinsfile

lankydan committed:
Merge pull request #6347 from corda/raz/os45to46

NOTICK: MERGE OS 4.5 -> OS 4.6

Commit:
CORDA-3578 add corda prefix to conf file names as original issue was … (#6322)

* CORDA-3578 add corda prefix to conf file names as original issue was having non-corda reference.conf files on classpath causes DriverDSLImp failure

it is better to have this naming convention and avoid further conflicts of conf files.

* fixed weird duplicates

* revert renaming changes for web-reference.conf and loadtest-reference.conf

lankydan committed:
CORDA-3841 Check `isAnyCheckpointPersisted` in `startFlowInternal` (#6351)

Only hit the database if `StateMachineState.isAnyCheckpointPersisted`
returns true. Otherwise, there will be no checkpoint to retrieve from the
database anyway. This can prevent errors due to a transient loss of
connection to the database.

adelel1 committed:
Merge remote-tracking branch 'origin/release/os/4.3' into adel/merge-4.3-to-4.4-17June

JamesHR3 committed:
EG-2654 - Ensure stack traces are printed to the logs in error reporting (#6345)

* EG-2654 Ensure stack trace is printed to the logs in error reporting

* EG-2654 - Add a test case for exception logging

chrisr3 committed:
CORDA-3831: Prevent CordappImpl TEST_INSTANCE crashing node when PWD is file-system root directory. (#6360)

woggioni committed:
removed duplicate gradle include (#6330)

subproject `tools:checkpoint-agent` is already included at line 63

chrisr3 committed:
ENT-5430: Fix deserialisation of commands containing generic types. (#6359)

rick-r3 committed:
Merge pull request #6358 from corda/adel/merge-4.3-to-4.4-17June

Merge OS 4.3 to OS 4.4

chrisr3 committed:
Merge commit 'fe617818895edab334d80c5e8de2b38f39e67af6' into chrisr3-os44-merge

mnesbit committed:
Merge pull request #6367 from corda/chrisr3-os44-merge

NOTICK: Merge from OS 4.4 -> OS 4.5 up to fe61781

Schife committed:
INFRA-394 enable full junit log to avoid truncation (#6368)

lankydan committed:
Merge branch 'release/os/4.5' into dan/os-4.5-to-4.6-merge-2020-06-18

# Conflicts:
#	node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt
#	node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt

lankydan committed:
CORDA-3841 Check `isAnyCheckpointPersisted` in `startFlowInternal` (#6351)

Only hit the database if `StateMachineState.isAnyCheckpointPersisted`
returns true. Otherwise, there will be no checkpoint to retrieve from the
database anyway. This can prevent errors due to a transient loss of
connection to the database.

Update tests after merging to 4.6

lankydan committed:
Merge pull request #6372 from corda/dan/os-4.5-to-4.6-merge-2020-06-18

NOTICK OS 4.5 to 4.6 merge 2020-06-18

wzur-r3 committed:
INFRA-347: Corda builds for Windows Servers (#6356)

cairncross committed:
INFRA-357 - Add release tag Jenkins build for OS 4.4 (#6362)

alicer3 committed:
center console message for registration (#6191)

Commit:
EG-1557 - Configuration data from "include" section ignored while command line contains the path to config file without leading ./ (#6354)

Configuration data from "include" section ignored while command line contains the path to config file without leading ./

wzur-r3 committed:
Merge branch 'release/os/4.4' into merge-44-45

wzur-r3 committed:
Merge pull request #6373 from corda/merge-44-45

NOTICK: merging OS 4.4 with OS 4.5

Schife committed:
INFRA-357 Adding JDK 11 Regression and Build tag jenkinsfile (#6374)

Schife committed:
NOTICK fix regression jenkinsfile typo (#6376)

wzur-r3 committed:
Remove obsolete Jenkins configration for smoke tests (#6379)

wzur-r3 committed:
Merge branch 'release/os/4.3' into merge-43-to-44

wzur-r3 committed:
Merge pull request #6380 from corda/merge-43-to-44

NOTICK: forward merge from Corda OS 4.3 to 4.4

wzur-r3 committed:
Merge branch 'release/os/4.4' into merge-44-to-45

wzur-r3 committed:
Merge pull request #6381 from corda/merge-44-to-45

NOTICK: merge Corda OS 4.4 to 4.5

wzur-r3 committed:
Merge branch 'release/os/4.5' into merge-45-to-46

wzur-r3 committed:
Merge pull request #6382 from corda/merge-45-to-46

NOTICK: Merge Corda OS 4.5 to 4.6

wzur-r3 committed:
Trying to ignore push events for Windows Server builds (#6383)

wzur-r3 committed:
Cosmetic change for Windows Server builds

Schife committed:
INFRA-357 Publish build tag  (#6375)

Schife committed:
Merge branch 'release/os/4.4' of https://github.com/corda/corda into INFRA-357

 Conflicts:
	testing/DockerfileJDK11Azul

Schife committed:
INFRA-357 Re adding JDK 11 regression build file that was removed in the 4.4 to 4.5 merge

wzur-r3 committed:
Merge pull request #6384 from corda/INFRA-357

INFRA-357 4.4 to 4.5 merge + JDK 11 regression build changes for tag publishing

wzur-r3 committed:
Merge pull request #6385 from corda/release/os/4.5

NOTICK Merge OS 4.5 -> OS 4.6

chrisr3 committed:
ENT-5430: Increase test coverage when serializing Optional fields. (#6387)

xavierlepretre committed:
Merge branch 'release/os/4.6' of github.com:xavierlepretre/corda into release/os/4.6

xavierlepretre committed:
Removed empty blocks.

nargas-ritu commented:
Internal backlog ticket [CORDA-4205](https://r3-cev.atlassian.net/browse/CORDA-4205)

marcellofarina75 commented:
Hello [xavierlepretre](https://github.com/xavierlepretre),
thank you for your contribution to Corda!

Our engineers reviewed your Pull Request and established that, while your proposal would definitely improve Corda, the work needed to merge it (in terms of unit, integration and regression tests, documentation updates, security assessments, and ongoing maintenance) would prevent us from shipping features and fixes that are prioritised higher in our roadmap (which is driven directly from inputs from across the user base).

So this means we regretfully need to let you know that your PR will be declined.

Please feel free to contact [devrel@r3.com](mailto:devrel@r3.com) if you need any further information
